# GATHER

In [1]:
import numpy as np
import pandas as pd
import zipfile
import dill

In [ ]:
#uncomment to save or load environment

dill.dump_session('notebook_session.db') #save
#dill.load_session('notebook_session.db') #load

In [126]:
zf = zipfile.ZipFile('./data/pisa2012.zip')
pisa2012 = pd.read_csv(zf.open('pisa2012.csv'), encoding = 'latin-1', sep = ',',
                       error_bad_lines = False, index_col = False, dtype = 'unicode', skipinitialspace=True)

MemoryError: 

> For our data wrangling process, I aim to focus on how both environmental factors and demographic factors affect a student's academic performance, future optimism and self-esteem. 

- Environmental factors will consist of ***teacher helpfulness***, ***sense of belonging***, and ***acculturation.***

- Demographic factors will consist of ***gender*** and ***parent background***.

- Academic performance will consist of math, reading, and science ***scores*** and ***work ethic***.
- ***Future optimism*** and ***self-esteem*** will both be standalone.

I'm essentially exploring the idea of, "If a student isn't getting any help and feels like an outsider, then that'll lower their commitment for school and hurt their self-esteem, in turn lowering their academic performance." On the flip side, a student who ***is*** getting help and feels comfortable in the school environment would have greater optimism in committing to school, thus raising their academic performance.

At the same time, I want to see how parent background, country, and gender can drive these variables against/in favor of the student.

After looking through the dataset, I have isolated these variables.

### Student Information
- **Gender + Parent Background**
    - **ST04Q01** - Gender
    - **PV1MATH** -	Plausible value 1 in mathematics
    - **PV1READ** - Plausible value 1 in reading
    - **PV1SCIE** -	Plausible value 1 in science
    - **OCOD1** ISCO-08 Occupation code - Mother
    - **OCOD2** ISCO-08 Occupation code - Father
    - **ST22Q01** -	Acculturation - Mother Immigrant (Filter)
    
- **STIDSTD** - student ID
- **CNT** - 3-character country code


### Teacher Helpfulness/Involvement
    
- **Teacher Support**
    - **ST83Q01** -	Teacher Support - Lets Us Know We Have to Work Hard
    - **ST83Q02** -	Teacher Support - Provides Extra Help When Needed
    - **ST83Q03** -	Teacher Support - Helps Students with Learning
    - **ST83Q04** -	Teacher Support - Gives Opportunity to Express Opinions
    
- **Student-Teacher Relation**
    - **ST86Q01** -	Student-Teacher Relation - Get Along with Teachers
    - **ST86Q02** -	Student-Teacher Relation - Teachers Are Interested
    - **ST86Q03** -	Student-Teacher Relation - Teachers Listen to Students
    - **ST86Q04** -	Student-Teacher Relation - Teachers Help Students
    - **ST86Q05** -	Student-Teacher Relation - Teachers Treat Students Fair
    
- **Teacher-Directed Instruction**
    - **ST79Q01** -	Teacher-Directed Instruction - Sets Clear Goals
    - **ST79Q02** -	Teacher-Directed Instruction - Encourages Thinking and Reasoning
    - **ST79Q06** -	Teacher-Directed Instruction - Checks Understanding
    - **ST79Q08** -	Teacher-Directed Instruction - Summarizes Previous Lessons
    - **ST79Q15** -	Teacher-Directed Instruction - Informs about Learning Goals
    
### Sense of Belonging

- **Sense of Belonging**
    - **ST87Q01** - Sense of Belonging - Feel Like Outsider
    - **ST87Q02** -	Sense of Belonging - Make Friends Easily
    - **ST87Q03** -	Sense of Belonging - Belong at School
    - **ST87Q04** -	Sense of Belonging - Feel Awkward at School
    - **ST87Q05** -	Sense of Belonging - Liked by Other Students
    - **ST87Q06** -	Sense of Belonging - Feel Lonely at School
    - **ST87Q07** -	Sense of Belonging - Feel Happy at School
    - **ST87Q08** -	Sense of Belonging - Things Are Ideal at School
    - **ST87Q09** -	Sense of Belonging - Satisfied at School
    - **BELONG**  - Sense of Belonging to School
    
### Acculturation

> The reason I picked only the host culture acculturations is because I want to see how well a student is ***fitting in*** to the culture of the school they're currently attending.

- **Acculturation**
    - **ST23Q01** -	Acculturation - Enjoy Host Culture Friends
    - **ST23Q03** -	Acculturation - Enjoy Host Culture Celebrations
    - **ST23Q05** -	Acculturation - Spend Time with Host Culture Friends
    - **ST23Q07** -	Acculturation - Participate in Host Culture Celebrations
    - **ST24Q01** -	Acculturation - Perceived Host-Heritage Cultural Differences - Values
    - **ST24Q02** -	Acculturation - Perceived Host-Heritage Cultural Differences - Mother Treatment
    - **ST24Q03** -	Acculturation - Perceived Host-Heritage Cultural Differences - Teacher Treatment


> I included math work ethic into overall work ethic. 
### Work Ethic

- **Math Work Ethic**
    - **ST46Q01** -	Math Work Ethic - Homework Completed in Time
    - **ST46Q02** -	Math Work Ethic - Work Hard on Homework
    - **ST46Q03** -	Math Work Ethic - Prepared for Exams
    - **ST46Q04** -	Math Work Ethic - Study Hard for Quizzes
    - **ST46Q05** -	Math Work Ethic - Study Until I Understand Everything
    - **ST46Q06** -	Math Work Ethic - Pay Attention in Classes
    - **ST46Q07** -	Math Work Ethic - Listen in Classes
    - **ST46Q08** -	Math Work Ethic - Avoid Distractions When Studying
    - **ST46Q09** -	Math Work Ethic - Keep Work Organized

- **Perseverance**
    - **ST93Q01** -	Perseverance - Give up easily
    - **ST93Q03** -	Perseverance - Put off difficult problems
    - **ST93Q04** -	Perseverance - Remain interested
    - **ST93Q06** -	Perseverance - Continue to perfection
    - **ST93Q07** -	Perseverance - Exceed expectations
    
- **Truancy**
    - **ST08Q01** - Truancy - Late for School
    - **ST09Q01** - Truancy - Skip whole school day
    
- **Out-of-School Study Time**
    - **ST57Q01** -	Out-of-School Study Time - Homework
    - **ST57Q02** -	Out-of-School Study Time - Guided Homework
    - **ST57Q03** -	Out-of-School Study Time - Personal Tutor
    - **ST57Q04** -	Out-of-School Study Time - Commercial Company
    - **ST57Q05** -	Out-of-School Study Time - With Parent
    - **ST57Q06** -	Out-of-School Study Time - Computer


### Future Optimism

- **Attitudes toward School**
    - **ST88Q01** -	Attitude towards School - Does Little to Prepare Me for Life
    - **ST88Q02** -	Attitude towards School - Waste of Time
    - **ST88Q03** -	Attitude towards School - Gave Me Confidence
    - **ST88Q04** -	Attitude towards School- Useful for Job
    - **ST89Q02** -	Attitude toward School - Helps to Get a Job
    - **ST89Q03** -	Attitude toward School - Prepare for College
    - **ST89Q04** -	Attitude toward School - Enjoy Good Grades
    - **ST89Q05** -	Attitude toward School - Trying Hard is Important
    
- **Instrumental Motivation**
    - **ST29Q07** -	Instrumental Motivation - Important for Future Study
    - **ST29Q08** -	Instrumental Motivation - Helps to Get a Job

### Self-Esteem
- **Perceived Control**
    - **ST91Q01** -	Perceived Control - Can Succeed with Enough Effort
    - **ST91Q02** -	Perceived Control - My Choice Whether I Will Be Good
    - **ST91Q03** -	Perceived Control - Problems Prevent from Putting Effort into School
    - **ST91Q04** -	Perceived Control - Different Teachers Would Make Me Try Harder
    - **ST91Q05** -	Perceived Control - Could Perform Well if I Wanted
    - **ST91Q06** -	Perceived Control - Perform Poor Regardless
    
- **Math Self-Concept** 
    - **ST42Q02** -	Math Self-Concept - Not Good at Maths
    - **ST42Q04** -	Math Self-Concept- Get Good Grades
    - **ST42Q06** -	Math Self-Concept - Learn Quickly
    - **ST42Q07** -	Math Self-Concept - One of Best Subjects
    - **ST42Q09** -	Math Self-Concept - Understand Difficult Work
    
After isolating the above variables, I intend to have these 14 columns in the final clean CSV:

- student ID
- country
- gender
- overall math score
- overall reading score
- overall science score
- teacher helpfulness/involvement
- sense of belonging
- acculturation
- work ethic
- future optimism
- self-esteem
- mother work background
- father work background

All variables from `teacher helpfulness` to `self-esteem` will be quantified with an average score, and normalized across all columns.

In [3]:
#create categories for each variable
student_info = ['STIDSTD', 'CNT', 'ST04Q01', 'PV1MATH', 'PV1READ', 'PV1SCIE', 'OCOD1', 'OCOD2', 'ST22Q01']

In [4]:
#teacher helpfulness/involvement
teacher_support = ['ST83Q01', 'ST83Q02', 'ST83Q03', 'ST83Q04'] 
teacher_relation = ['ST86Q01', 'ST86Q02', 'ST86Q03', 'ST86Q04', 'ST86Q05']
teacher_direct = ['ST79Q01', 'ST79Q02', 'ST79Q06', 'ST79Q08', 'ST79Q15']

teacher_helpfulness = np.concatenate([teacher_support, teacher_relation, teacher_direct]).ravel()

#num of columns
len(pisa2012[teacher_helpfulness].columns)

14

In [5]:
#sense of belonging
belonging = ['ST87Q01', 'ST87Q02', 'ST87Q03', 'ST87Q04', 'ST87Q05', 'ST87Q06', 'ST87Q07', 'ST87Q08', 'ST87Q09'] 

len(pisa2012[belonging].columns)

9

In [6]:
#acculturation
acculturation = ['ST23Q01', 'ST23Q03', 'ST23Q05', 'ST23Q07', 'ST24Q01', 'ST24Q02', 'ST24Q03']

len(pisa2012[acculturation].columns)

7

In [7]:
#work ethic
math_ethic = ['ST46Q01', 'ST46Q02', 'ST46Q03', 'ST46Q04', 'ST46Q05', 'ST46Q06', 'ST46Q07', 'ST46Q08', 'ST46Q09']
perseverance = ['ST93Q01', 'ST93Q03', 'ST93Q04', 'ST93Q06', 'ST93Q07']
truancy = ['ST08Q01', 'ST09Q01']

work_ethic = np.concatenate([math_ethic, perseverance, truancy]).ravel()
pisa2012[work_ethic].columns.shape[0]

16

In [8]:
#future optimism
attitude_school =['ST88Q01', 'ST88Q02', 'ST88Q03', 'ST88Q04', 'ST89Q02', 'ST89Q03', 'ST89Q04', 'ST89Q05']
motivation = ['ST29Q07', 'ST29Q08']

future_optimism = np.concatenate([attitude_school, motivation]).ravel()
pisa2012[future_optimism].columns.shape[0]

10

In [9]:
#self esteem
perceived_control = ['ST91Q01', 'ST91Q02', 'ST91Q03', 'ST91Q04', 'ST91Q05', 'ST91Q06']
math_self_concept = ['ST42Q02', 'ST42Q04', 'ST42Q06', 'ST42Q07', 'ST42Q09']

self_esteem = np.concatenate([perceived_control, math_self_concept]).ravel()
pisa2012[self_esteem].columns.shape[0]

11

### Observations
1. `teacher_helpfulness` has a total of 14 columns.
2. `belonging` has a total of 9 columns.
3. `acculturation` has a total of 7 columns.
4. `work ethic` has a total of 16 columns.
5. `future_optimism` has a total of 10 columns.
6. `self_esteem` has a total of 11 columns.
7. `student info` has a total of 9 columns.


All variables total up to 76 columns. We'll reduce those 76 columns to only 15 columns (observations 1 to 6, plus all `student_info` columns)

# ASSESS

### Overall

In [10]:
pisa2012.shape

(485490, 636)

> There are 485490 entries and 636 columns. We have some **serious** cleaning to do.

In [11]:
#number of rows with duplicate data
num_duplicate = pisa2012.duplicated().sum()
num_duplicate

0

In [13]:
#count how many rows have missing data
num_missing = pisa2012.isna().sum().max()
num_missing

472209

In [14]:
print('This PISA dataset has {} rows with no missing data.'.format(pisa2012.shape[0] - num_missing))

This PISA dataset has 13281 rows with no missing data.


> Comb through each variable category one at a time.

### Student Info

In [15]:
df_student_info = pisa2012[student_info]

In [16]:
df_student_info.head()

STIDSTD      CNT ST04Q01   PV1MATH   PV1READ   PV1SCIE  \
0   00001  Albania  Female  406.8469  249.5762  341.7009   
1   00002  Albania  Female  486.1427  406.2936  548.9929   
2   00003  Albania  Female  533.2684    401.21  499.6643   
3   00004  Albania  Female  412.2215   547.363  438.6796   
4   00005  Albania  Female  381.9209  311.7707  361.5628   

                                        OCOD1  \
0                         Building architects   
1  Tailors, dressmakers, furriers and hatters   
2                                   Housewife   
3                                   Housewife   
4                                   Housewife   

                                               OCOD2 ST22Q01  
0                            Primary school teachers     NaN  
1                    Building construction labourers     NaN  
2                    Bricklayers and related workers     NaN  
3  Cleaners and helpers in offices, hotels and ot...     NaN  
4                                         Economists     NaN

In [26]:
df_student_info.describe()

STIDSTD     CNT ST04Q01  PV1MATH   PV1READ   PV1SCIE      OCOD1  \
count   485490  485490  485490   485490    485490    485490     483887   
unique   33806      68       2     7063     11556      6153        588   
top      00200  Mexico  Female  444.158  510.1099  472.5289  Housewife   
freq        68   33806  245064      359       355       753      74358   

          OCOD2 ST22Q01  
count    482936   40721  
unique      588       2  
top     Missing     Yes  
freq      36559   31692

In [27]:
#first 3 results of most common father occupations
pisa2012[student_info].OCOD2.value_counts()[:3]

Missing                                                         36559
Vague(a good job, a quiet job, a well paid job, an office jo    14716
Heavy truck and lorry drivers                                   11816
Name: OCOD2, dtype: int64

- It appears that Mexico is the most common country, with a frequency of 33806 out of 485490 total students.
- It appears that there are more female than male students.
- It appears that the most common occupation for a mother is a housewife.
- Besides missing and vague information, it appears that the most common occupation for a father is a heavy truck/lorry driver.
- It appears that most student's mothers is an immigrant.


- It appears that the most common math score is 444.158, with a frequency of 359 out of 485490 total scores. 
- It appears that the most common reading score is 510.1099, with a frequency of 355 out of 485490 total scores.
- It appears that the most common science score is 472.5289, with a frequency of 588 out of 485490 total scores.

In [28]:
df_student_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 9 columns):
STIDSTD    485490 non-null object
CNT        485490 non-null object
ST04Q01    485490 non-null object
PV1MATH    485490 non-null object
PV1READ    485490 non-null object
PV1SCIE    485490 non-null object
OCOD1      483887 non-null object
OCOD2      482936 non-null object
ST22Q01    40721 non-null object
dtypes: object(9)
memory usage: 33.3+ MB


**Note:** math, reading, and science scores are recorded as strings.

**Student ID**

In [29]:
#check for number of unique student IDs
df_student_info.STIDSTD.duplicated().sum()

451684

There are 451684 duplicated student IDs, but that does not mean every student is the same. If we take a look below at some of the repeated student IDs, we can see that they are mostly based in different countries, all with much different records for the other pieces of student information.

In [30]:
df_student_info[df_student_info['STIDSTD'].duplicated()].sample(10)

STIDSTD              CNT ST04Q01   PV1MATH   PV1READ   PV1SCIE  \
463259   02277          Tunisia    Male    282.84  236.4753  265.5166   
435459   02454  Slovak Republic    Male  755.8109  664.7942  682.0589   
229980   02970          Ireland  Female  524.3885  619.7247  617.0645   
132254   04481   Czech Republic    Male  330.4331  308.6504  393.9202   
86944    12475           Canada  Female  501.4878  552.2083  547.5009   
325772   10070           Mexico    Male  501.4878  493.8195  496.7736   
373223   00337             Peru    Male  323.4227  312.9007  334.5207   
112157   04915            Chile  Female  379.1946  423.3712  356.9004   
102290   06277      Switzerland    Male  466.7472  517.2363  456.6766   
125580   02409       Costa Rica    Male  443.8464  438.4853   468.799   

                                           OCOD1  \
463259                                 Housewife   
435459                  Applications programmers   
229980                     Nursing professionals   
132254                               Do not know   
86944   University and higher education teachers   
325772             Domestic cleaners and helpers   
373223                                   Waiters   
112157        Plastic products machine operators   
102290                 Laundry machine operators   
125580                                 Housewife   

                                                    OCOD2 ST22Q01  
463259                                       Shop keepers     NaN  
435459  Business services and administration managers ...     NaN  
229980  Cleaners and helpers in offices, hotels and ot...     NaN  
132254  Vague(a good job, a quiet job, a well paid job...     NaN  
86944                    Musicians, singers and composers     NaN  
325772                                   Freight handlers     NaN  
373223                      Subsistence livestock farmers     NaN  
112157                          Plumbers and pipe fitters     NaN  
102290  Social beneficiary (unemployed, retired, sickn...     NaN  
125580                 Electrical engineering technicians     NaN

Let's pick 1 duplicated student ID and examine it.

In [31]:
df_student_info[df_student_info['STIDSTD'] == '03425'].sample(10)

STIDSTD             CNT ST04Q01   PV1MATH   PV1READ   PV1SCIE  \
25575    03425       Australia  Female  419.3878  476.5106  438.2134   
117523   03425        Colombia  Female   249.112  363.9568  300.0186   
275087   03425          Jordan    Male   364.239  344.9785  363.0548   
19667    03425       Argentina    Male  365.7969  340.9688  344.4051   
464407   03425         Tunisia    Male  523.7654  476.8183   539.295   
187928   03425          France  Female   352.477  317.3308  341.9806   
468814   03425          Turkey  Female  493.3089  524.4075   487.169   
131198   03425  Czech Republic  Female  374.8325  417.9699  407.2548   
220003   03425         Hungary  Female  377.2472  426.7073  398.2096   
403641   03425  China-Shanghai  Female  695.8327   610.193  581.3502   

                                           OCOD1  \
25575                              Sales workers   
117523           Beauticians and related workers   
275087                                 Housewife   
19667                                    Missing   
464407                                 Housewife   
187928                Cashiers and ticket clerks   
468814                                 Housewife   
131198                                   Waiters   
220003                               Do not know   
403641  University and higher education teachers   

                                                    OCOD2 ST22Q01  
25575                               Construction managers     NaN  
117523                                                NaN     NaN  
275087  Social beneficiary (unemployed, retired, sickn...     NaN  
19667                                             Missing     NaN  
464407                                     House builders     NaN  
187928                                        Do not know     NaN  
468814                            Religious professionals     NaN  
131198                          Car, taxi and van drivers     NaN  
220003                     Toolmakers and related workers     NaN  
403641                                   Finance managers     NaN

This piece of evidence, coupled with the fact there are ***no*** duplicated entries for the entire PISA dataset, proves that repeated student IDs does not mean repeated identical students.

These repeated student IDs are most likely caused by other factors such as ***different schools*** among different countries. Additionally, students with the same student ID all have different records--some are recorded as male and other females, based in different countries, and with varying math scores.

**Country**

In [32]:
#number of entries with missing countries
sum(df_student_info.CNT.isnull())

0

In [33]:
df_student_info.CNT.nunique()

68

> Our dataset includes data from 65 countries, but our DataFrame shows that there are 68 countries. What are the 3 additional rogue countries then?

In [34]:
#Massachusetts, Florida, and Connecticut recorded as separate countries.
df_student_info.CNT.value_counts()[-5:]

Florida (USA)               1896
Perm(Russian Federation)    1761
Massachusetts (USA)         1723
Connecticut (USA)           1697
Liechtenstein                293
Name: CNT, dtype: int64

**Gender**

In [35]:
#number of entries with no gender entry
sum(df_student_info.ST04Q01.isnull())

0

In [36]:
#check male to female ratio
gender_counts = df_student_info.ST04Q01.value_counts()
gender_counts

Female    245064
Male      240426
Name: ST04Q01, dtype: int64

In [37]:
total_students = gender_counts[0] + gender_counts[1]
print('Female students make up {:0.2f}% of the student population.'.format((gender_counts[0] / total_students) * 100))
print('Male students make up {:0.2f}% of the student population.'.format((gender_counts[1] / total_students) * 100))

Female students make up 50.48% of the student population.
Male students make up 49.52% of the student population.


> Male and female students are about **evenly split** with female students outnumbering male students by only .96%.

**Math, Reading, and Science Score**

> All scores are recorded as strings. To appropriately assess the numbers quantitatively, I will temporarily change them to floats.

In [38]:
#number of entries with no entry for math, science, and reading scores
print(sum(df_student_info.PV1MATH.isnull()))
print(sum(df_student_info.PV1READ.isnull()))
print(sum(df_student_info.PV1SCIE.isnull()))

0
0
0


In [39]:
#look at 5-Number Summary for all scores
df_student_info[['PV1MATH', 'PV1READ', 'PV1SCIE']].astype(float).describe()

PV1MATH        PV1READ        PV1SCIE
count  485490.000000  485490.000000  485490.000000
mean      469.621653     472.004640     475.769824
std       103.265391     102.505523     101.464426
min        19.792800       0.083400       2.648300
25%       395.318600     403.600700     404.457300
50%       466.201900     475.455000     475.699400
75%       541.057800     544.502500     547.780700
max       962.229300     904.802600     903.338300

- It appears that people, on average, scored higher for science.
- All people scored similarly in the 25th, 50th, and 75th percentiles for math, reading, and science scores.
- The highest recorded score is for math at 962.229.
- The lowest recorded score is for reading at 0.083400.

> Initially, I was going to handle missing and duplicate data among the upcoming **environmental + student effect** variables. However, I concluded that it's unncecessary to drop rows with missing/duplicate data. This is because we have previously established that 1) there are no duplicate records of the same students, and 2) NaN data is to be expected especially under columns that are not required information. As long as our student IDs, scores, and gender (among other necessary column fields) equal the entire number of students, then we have nothing to worry about.

In that case, the rest of the data assessment focuses only on 1) isolating variables, 2) finding unique values, and 3) finding a way to normalize them quantitatively.

In [40]:
#function to isolate unique values among multiple columns in a df
def comb_cols(df, columns):
    return (pd.unique(df[columns].values.ravel()))

### Teacher Helpfulness/Involvement

-**Teacher Helpfulness** measures how well a teacher is dedicating their time to helping a student drive their academic performance.


In [41]:
#unique values across all teacher support categories
pd.unique(pisa2012[teacher_helpfulness].values.ravel())

array(['Strongly disagree', 'Agree', 'Disagree', 'Never or Hardly Ever',
       'Most Lessons', 'Every Lesson', nan, 'Strongly agree',
       'Some Lessons'], dtype=object)

> It appears some categories under `teacher_helpfulness` consist of different values. Let's check them more in-depth:

In [42]:
#list all unique values per column under teacher_helpfulness
print('Teacher support:', comb_cols(pisa2012, teacher_support))
print('Teacher relation:', comb_cols(pisa2012, teacher_relation))
print('Teacher direction:', comb_cols(pisa2012, teacher_direct))

Teacher support: ['Strongly disagree' 'Agree' nan 'Strongly agree' 'Disagree']
Teacher relation: ['Agree' 'Strongly disagree' 'Disagree' nan 'Strongly agree']
Teacher direction: ['Never or Hardly Ever' 'Most Lessons' 'Every Lesson' nan 'Some Lessons']


- Variables among `teacher_support` and `teacher_relation` are the exact same at **"Strongly disagree," "Disagree," "Agree," "Strongly agree," and "NaN."** Therefore, we don't have to worry about future quantification and normalization between these 2 variables.

- However, variables under `teacher_direct` are ***"Never or Hardly Ever," "Most Lessons," "Every Lesson," "Some Lessons," and "NaN."*** 

### Sense of Belonging

- **Sense of Belonging** measures how comfortable a student feels in their immediate social environment at school.

In [43]:
#list all unique values per column
print('Sense of Belonging:', comb_cols(pisa2012, belonging))

Sense of Belonging: ['Agree' 'Strongly disagree' 'Strongly agree' nan 'Disagree']


- `belonging` has the same column values as `teacher_support` and `teacher_relation`.

### Acculturation
- **Acculturation** measures how well a student is assimilating into their host culture.

In [44]:
print('Acculturation:', comb_cols(pisa2012, acculturation))

Acculturation: [nan 'Agree' 'Disagree' 'Strongly disagree' 'Strongly agree']


- `acculturation` has the same column values as `belonging`, `teacher_support`, and `teacher_relation`.

### Work Ethic

- **Math Work Ethic** measures a student's focus and dedication to learning math.
- **Perseverance** measures a student's ability to not give up and do better on their work.
- **Truancy** measures a student's attendance.

In [45]:
#comb unique values through all columns
print(comb_cols(pisa2012, work_ethic))

['Agree' 'Very much like me' 'Somewhat like me' 'None  ' '1'
 'Strongly agree' 'Disagree' 'Not at all like me' 'Mostly like me'
 'One or two times  ' 'Not much like me' nan '2' 'Three or four times  '
 '3' 'Strongly disagree' 'Five or more times  ' '4']


In [46]:
#comb unique values individually
print('Math Ethic:', comb_cols(pisa2012, math_ethic))
print('Perseverance:', comb_cols(pisa2012, perseverance))
print('Truancy:', comb_cols(pisa2012, truancy))

Math Ethic: ['Agree' 'Strongly agree' 'Disagree' nan 'Strongly disagree']
Perseverance: ['Very much like me' 'Somewhat like me' 'Not at all like me'
 'Mostly like me' 'Not much like me' nan]
Truancy: ['None  ' '1' 'One or two times  ' '2' 'Three or four times  ' nan '3'
 'Five or more times  ' '4']


- `math_ethic` has the same column values as `acculturation`, `belonging`, `teacher_support`, and `teacher_relation`.
- `perseverance` and `truancy` have no common values with each other or any other previous variable.

### Future Optimism
- **Attitude towards School** measures a student's current perception of how useful school is in preparing them for college or a job.
- **Instrumental Motivation** measures a student's current perception of how helpful school is in future career/study prospects.

In [47]:
#comb unique values through all columns
comb_cols(pisa2012, future_optimism)

array(['Agree', 'Strongly agree', 'Strongly disagree', nan, 'Disagree'],
      dtype=object)

> Both columns under `future_optimism` share the same values. But just to double-check, let's check them individually:

In [48]:
print('Attitudes toward School: ', comb_cols(pisa2012, attitude_school))
print('Motivation: ', comb_cols(pisa2012, motivation))

Attitudes toward School:  ['Agree' 'Strongly agree' 'Strongly disagree' nan 'Disagree']
Motivation:  ['Agree' 'Strongly agree' 'Disagree' nan 'Strongly disagree']


- both `attitude_school` and `motivation` have the same values in common with `math_ethic`, `acculturation`, `belonging`, `teacher_support`, and `teacher_relation`.

### Self-Esteem
- **Perceived Control** measures if a student feels in control of their success.
- **Math Self-Concept** measures how a student feels about themselves in relation to math.

In [49]:
#comb unique values through all columns
comb_cols(pisa2012, self_esteem)

array(['Strongly agree', 'Strongly disagree', 'Disagree', 'Agree', nan],
      dtype=object)

In [50]:
#comb unique values individually
print('Perceived Control: ', comb_cols(pisa2012, perceived_control))
print('Math Self-Concept: ', comb_cols(pisa2012, math_self_concept))

Perceived Control:  ['Strongly agree' 'Strongly disagree' 'Disagree' nan 'Agree']
Math Self-Concept:  ['Disagree' 'Agree' nan 'Strongly agree' 'Strongly disagree']




- Values among `teacher_support`, `teacher_relation`, `belonging`, `acculturation`, `math_ethic`, `attitude_school`, `motivation`, `perceived_control`, and `math_self_concept` are **"Strongly disagree," "Disagree," "Agree," "Strongly agree," and "NaN."**


- Values under `teacher_direct` are **"Every Lesson", "Most Lessons", "Some Lessons", "Never or Hardly Ever", and "NaN."**



- Values under `perseverance` are **"Very much like me," "Somewhat like me," "Not at all like me," "Mostly like me," "Not much like me," and "NaN."** 


- Values under `truancy` are **"Five or more times," "4", "Three or four times", "3", "2", "One or Two Times", "1", "None", and "NaN."**

## Quality
- Erroneous data types for math, reading, and science scores.
- Florida (USA), Massachusetts (USA), and Connecticut (USA) considered as separate countries from United States of America.
- Ordinal nature of categorical variables among these variable groups means they can be quantified:
    - `teacher_support`, `teacher_relation`, `belonging`, `acculturation`, `math_ethic`, `attitude_school`, `motivation`, `perceived_control`, `math_self_concept`
    - `teacher_direct`
    - `perseverance`
    - `truancy`
- Unnormalized values among previous variable groups.
- Under `mother_occupation` and `father_occupation`, 'Missing' is in place of 'NaN.'
- Student ID not explicitly defined as a string.

## Tidiness
- Abundance of unnecessary variables besides the ones isolated in pre-assessment stage.

# CLEAN

In [43]:
pisa2012_clean = pisa2012.copy()

## Quality
> **ISSUE 1:** Erroneous data types for math, reading, and science scores.

**Define:** convert `PV1MATH`, `PV1READ`, and `PV1SCIE` columns to float data type.

### Code

In [44]:
#isolate score columns
scores = ['PV1MATH', 'PV1READ', 'PV1SCIE']
pisa2012_clean[scores].head()

PV1MATH   PV1READ   PV1SCIE
0  406.8469  249.5762  341.7009
1  486.1427  406.2936  548.9929
2  533.2684    401.21  499.6643
3  412.2215   547.363  438.6796
4  381.9209  311.7707  361.5628

In [45]:
#convert columns to float
pisa2012_clean[scores] = pisa2012_clean[scores].astype(float)

### Test

In [46]:
pisa2012_clean[scores].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 3 columns):
PV1MATH    485490 non-null float64
PV1READ    485490 non-null float64
PV1SCIE    485490 non-null float64
dtypes: float64(3)
memory usage: 11.1 MB


> **ISSUE 2:** Florida (USA), Massachusetts (USA), and Connecticut (USA) considered as separate countries from United States of America.

**Define:** change state names to United States of America.

### Code

In [47]:
#isolate state names
states = ['Florida (USA)', 'Massachusetts (USA)', 'Connecticut (USA)']

In [48]:
#correct states
pisa2012_clean.loc[pisa2012_clean.CNT.isin(states), 'CNT'] = 'United States of America'

### Test

In [49]:
#there should be only 65 countries instead of 68
pisa2012_clean.CNT.nunique()

65

> **ISSUE 3:** 
- Ordinal nature of categorical variables among these variable groups means they can be quantified:
    1. `teacher_support`, `teacher_relation`, `belonging`, `acculturation`, `math_ethic`, `attitude_school`, `motivation`, `perceived_control`, `math_self_concept`
    2. `teacher_direct`
    3. `perseverance`
    4. `truancy`

**Define:** Come up with a scoring system for each variable group.

> For each variable group, each value will be assigned a value as follows:

**Group 1**
- "Strongly agree" : 4
- "Agree" : 3
- "Disagree" : 2
- "Strongly Disagree" : 1

**Group 2 (Teacher-Directed Instruction)**
- "Every Lesson" : 4
- "Most Lessons" : 3
- "Some Lessons" : 2
- "Never or Hardly Ever": 1

**Group 3 (Perseverance)**
- "Very much like me" : 5
- "Mostly like me" : 4
- "Somewhat like me" : 3
- "Not much like me" : 2
- "Not at all like me" : 1

**Group 4 (Truancy)**
- "Five or more times" : 5
- "Three or four times" : 3
- "One or Two Times" : 1

***Before going on to create scoring systems, it is important to keep note of the following:***
- under `belonging`, ST87Q01, ST87Q04, ST87Q06 ask the student ***negative feelings*** about themselves. 
- under `attitude_school`, ST88Q01 and ST88Q02 evaluate negative ***perceptions*** the student has about school.
- under `perceived_control`, ST91Q03 and ST91Q06 evaluate things the student feels is ***out of their control***.
- under `math_self_concept`, ST42Q02 focuses on ***pessimistic self-perception***.

- under `perseverance`, ST93Q01 and ST93Q03 evaluate ***demotivating factors*** against the student's work ethic.

> For the above individual questions, ***the scoring values will be reversed.*** For example, "Strongly agree" will be assigned 1 instead of 4. And so forth.

### Code

In [50]:
#isolate reverse questions

#group1
belonging_reverse = ['ST87Q01', 'ST87Q04', 'ST87Q06']
attitude_school_reverse = ['ST88Q01', 'ST88Q02']
perceived_control_reverse = ['ST91Q03', 'ST91Q06']
math_self_concept_reverse = ['ST42Q02']

#group3
perseverance_reverse = np.array(['ST93Q01', 'ST93Q03'])

In [51]:
#group questions to be reversed
group1_reverse = np.concatenate([belonging_reverse, attitude_school_reverse,
                                 perceived_control_reverse, math_self_concept_reverse]).ravel()
group3_reverse = perseverance_reverse

In [52]:
#isolate all question groups 
group1 = (np.concatenate([teacher_support, teacher_relation, belonging, acculturation, math_ethic,
          attitude_school, motivation, perceived_control, math_self_concept]).ravel())
group2 = teacher_direct
group3 = perseverance
group4 = truancy

> Even though variable groups 2 through 4 are singular fields, I assigned them to new variables anyway for the sake of code readability.

I want to deal with score assignment on **normal questions** and **reverse-scoring questions** separately. This means getting rid of the reverse-scoring questions from the main grouped variables.

In [53]:
#index to-be-reversed questions from group 1 and 3
indices1 = [i for (i, v) in enumerate(group1) if (v == group1_reverse).any()]
indices3 = [i for (i, v) in enumerate(group3) if (v == group3_reverse).any()]

In [54]:
#remove to-be-reversed questions from group 1 and 3
group1_clean = np.delete(group1, indices1)
group3_clean = np.delete(group3, indices3)

### Test

> Compare our original `pisa2012` DataFrame to our `pisa2012_clean` one, since the clean version has been altered

In [55]:
#verify we separated our group 1 normal and to-be-reversed scores correctly
print('The original dataset has {} columns for group 1.'.format(pisa2012[group1].shape[1]))
print('There are {} columns whose scores we need to reverse.'.format(pisa2012[group1_reverse].shape[1]))
print('Group 1 for our dataset now has {} columns.'.format(pisa2012_clean[group1_clean].shape[1]))

The original dataset has 55 columns for group 1.
There are 8 columns whose scores we need to reverse.
Group 1 for our dataset now has 47 columns.


In [56]:
#verify we separated our group 3 scores correctly
print('The original dataset has {} columns for group 3.'.format(pisa2012[group3].shape[1]))
print('There are {} columns whose scores we need to reverse.'.format(pisa2012[group3_reverse].shape[1]))
print('Group 3 for our dataset now has {} columns.'.format(pisa2012_clean[group3_clean].shape[1]))

The original dataset has 5 columns for group 3.
There are 2 columns whose scores we need to reverse.
Group 3 for our dataset now has 3 columns.


### Code

**Group 1 Score Assignment**

In [57]:
#create score dictionary
group1_score = {'Strongly agree' : 4,
               'Agree' : 3,
               'Disagree' : 2,
               'Strongly disagree' : 1}

In [58]:
#reverse score dictionary
group1_score_reverse = {'Strongly agree' : 1,
                       'Agree' : 2,
                       'Disagree' : 3,
                       'Strongly disagree' : 4}

In [59]:
#assign reverse scoring first
pisa2012_clean[group1_reverse] = pisa2012_clean[group1_reverse].stack().map(group1_score_reverse).unstack()

In [60]:
#assign normal scoring
pisa2012_clean[group1_clean] = pisa2012_clean[group1_clean].stack().map(group1_score).unstack()

### Test

- Compare both original and clean versions of DataFrame, as our clean version is altered.

In [61]:
#make sure counts are consistent 

#reverse questions
pisa2012[group1_reverse].apply(pd.Series.value_counts).reindex(index = ['Strongly agree', 'Agree', 'Disagree', 'Strongly disagree'])

ST87Q01  ST87Q04  ST87Q06  ST88Q01  ST88Q02  ST91Q03  \
Strongly agree       13844    12453    10164    30700    11433    29721   
Agree                27910    33872    25413    74493    25060    83815   
Disagree            137026   137745   125801   151993   154332   133990   
Strongly disagree   132996   126928   149574    54064   120139    62300   

                   ST91Q06  ST42Q02  
Strongly agree       19941    51409  
Agree                47525    91811  
Disagree            139579   118658  
Strongly disagree   102611    51624

In [62]:
pisa2012_clean[group1_reverse].apply(pd.Series.value_counts)

ST87Q01  ST87Q04  ST87Q06  ST88Q01  ST88Q02  ST91Q03  ST91Q06  ST42Q02
1.0    13844    12453    10164    30700    11433    29721    19941    51409
2.0    27910    33872    25413    74493    25060    83815    47525    91811
3.0   137026   137745   125801   151993   154332   133990   139579   118658
4.0   132996   126928   149574    54064   120139    62300   102611    51624

In [65]:
#normal questions
pisa2012[group1_clean].apply(pd.Series.value_counts).reindex(index = ['Strongly disagree', 'Disagree', 'Agree', 'Strongly agree'])

ST83Q01  ST83Q02  ST83Q03  ST83Q04  ST86Q01  ST86Q02  \
Strongly disagree     7685    12929    10361    18256     7036     9483   
Disagree             30990    38084    32368    43660    43810    51359   
Agree               161908   148382   152294   147403   180918   175300   
Strongly agree      112922   113717   117920   103626    81459    76449   

                   ST86Q03  ST86Q04  ST86Q05  ST87Q02  ...  ST29Q07  ST29Q08  \
Strongly disagree    11629     9991    12589     6134  ...    26612    21700   
Disagree             62855    42840    40843    31913  ...    65853    57949   
Agree               168284   173260   174259   174067  ...   132131   150347   
Strongly agree       69420    86203    84213   100024  ...    90470    85236   

                   ST91Q01  ST91Q02  ST91Q04  ST91Q05  ST42Q04  ST42Q06  \
Strongly disagree     2010     6201    38360     6493    28895    35775   
Disagree              8977    38802   137355    26872    96860   111521   
Agree               122392   133755    89947   143900   135358   119008   
Strongly agree      178051   131638    43736   132345    50867    46023   

                   ST42Q07  ST42Q09  
Strongly disagree    74462    57153  
Disagree            110599   130944  
Agree                77922    93920  
Strongly agree       49600    30206  

[4 rows x 47 columns]

In [64]:
pisa2012_clean[group1_clean].apply(pd.Series.value_counts)

ST83Q01  ST83Q02  ST83Q03  ST83Q04  ST86Q01  ST86Q02  ST86Q03  ST86Q04  \
1.0     7685    12929    10361    18256     7036     9483    11629     9991   
2.0    30990    38084    32368    43660    43810    51359    62855    42840   
3.0   161908   148382   152294   147403   180918   175300   168284   173260   
4.0   112922   113717   117920   103626    81459    76449    69420    86203   

     ST86Q05  ST87Q02  ...  ST29Q07  ST29Q08  ST91Q01  ST91Q02  ST91Q04  \
1.0    12589     6134  ...    26612    21700     2010     6201    38360   
2.0    40843    31913  ...    65853    57949     8977    38802   137355   
3.0   174259   174067  ...   132131   150347   122392   133755    89947   
4.0    84213   100024  ...    90470    85236   178051   131638    43736   

     ST91Q05  ST42Q04  ST42Q06  ST42Q07  ST42Q09  
1.0     6493    28895    35775    74462    57153  
2.0    26872    96860   111521   110599   130944  
3.0   143900   135358   119008    77922    93920  
4.0   132345    50867    46023    49600    30206  

[4 rows x 47 columns]

### Code

**Group 2 Score Assignment**

In [66]:
#create score dictionary
group2_score = {'Every Lesson' : 4,
                'Most Lessons' : 3,
                'Some Lessons' : 2,
                'Never or Hardly Ever': 1}

In [67]:
#assign scoring
pisa2012_clean[group2] = pisa2012_clean[group2].stack().map(group2_score).unstack()

### Test

In [68]:
#make sure counts are consistent
pisa2012[group2].apply(pd.Series.value_counts).reindex(index = ['Never or Hardly Ever', 'Some Lessons', 'Most Lessons', 'Every Lesson'])

ST79Q01  ST79Q02  ST79Q06  ST79Q08  ST79Q15
Never or Hardly Ever    19716    44116    18984    62461    14020
Some Lessons            64800    92573    60052    98976    51228
Most Lessons           116036   104082   104426    79561   107336
Every Lesson           114357    73557   130413    73203   141262

In [69]:
pisa2012_clean[group2].apply(pd.Series.value_counts)

ST79Q01  ST79Q02  ST79Q06  ST79Q08  ST79Q15
1.0    19716    44116    18984    62461    14020
2.0    64800    92573    60052    98976    51228
3.0   116036   104082   104426    79561   107336
4.0   114357    73557   130413    73203   141262

### Code

**Group 3 Score Assignment**

In [70]:
#create score dictionary
group3_score = {'Very much like me' : 5,
                'Mostly like me' : 4,
                'Somewhat like me' : 3,
                'Not much like me' : 2,
                'Not at all like me' : 1}

In [71]:
#create reverse score dictionary
group3_score_reverse = {'Very much like me' : 1,
                'Mostly like me' : 2,
                'Somewhat like me' : 3,
                'Not much like me' : 4,
                'Not at all like me' : 5}

In [72]:
#assign reverse scoring
pisa2012_clean[group3_reverse] = pisa2012_clean[group3_reverse].stack().map(group3_score_reverse).unstack()

In [73]:
#assign normal scoring
pisa2012_clean[group3_clean] = pisa2012_clean[group3_clean].stack().map(group3_score).unstack()

### Test

In [74]:
#make sure counts are consistent

#reverse question scores
pisa2012[group3_reverse].apply(pd.Series.value_counts).reindex(index = ['Very much like me', 'Mostly like me', 'Somewhat like me', 'Not much like me', 'Not at all like me'])

ST93Q01  ST93Q03
Very much like me     22791    34481
Mostly like me        34137    59230
Somewhat like me      78831    96460
Not much like me     109281    79384
Not at all like me    67816    42585

In [75]:
pisa2012_clean[group3_reverse].apply(pd.Series.value_counts)

ST93Q01  ST93Q03
1.0    22791    34481
2.0    34137    59230
3.0    78831    96460
4.0   109281    79384
5.0    67816    42585

In [76]:
#normal group 3 questions
pisa2012[group3_clean].apply(pd.Series.value_counts).reindex(index = ['Not at all like me', 'Not much like me', 'Somewhat like me', 'Mostly like me', 'Very much like me'])

ST93Q04  ST93Q06  ST93Q07
Not at all like me    12077    16112    20853
Not much like me      39506    51802    59466
Somewhat like me      90208    88682    96188
Mostly like me       104572    88555    78451
Very much like me     64948    67119    57301

In [77]:
pisa2012_clean[group3_clean].apply(pd.Series.value_counts)

ST93Q04  ST93Q06  ST93Q07
1.0    12077    16112    20853
2.0    39506    51802    59466
3.0    90208    88682    96188
4.0   104572    88555    78451
5.0    64948    67119    57301

And finally . . . for group 4! 

### Code
**Group 4 Score Assignment**

In [78]:
#create score dictionary
group4_score = {'Five or more times' : 5, 
                'Three or four times' : 3,
                'One or two times' : 1}

In [79]:
pd.unique(pisa2012_clean[group4].values.ravel())

array(['None  ', 'One or two times  ', 'Three or four times  ', nan,
       'Five or more times  '], dtype=object)

**Unless I get rid of white space under Group 4 values, my scoring assignment won't work since the current answers won't be recognized in the `group4_score` dictionary.**

In [80]:
#strip all white space from group 4 values
pisa2012_clean[group4] = pisa2012_clean[group4].apply(lambda x: x.str.strip())

In [81]:
#assign scoring
pisa2012_clean[group4] = pisa2012_clean[group4].stack().map(group4_score).unstack()

### Test

In [82]:
pisa2012[group4].apply(pd.Series.value_counts)[1:]

ST08Q01  ST09Q01
One or two times        124380    75969
Three or four times      29817    10882
Five or more times       18881     6282

In [83]:
pisa2012_clean[group4].apply(pd.Series.value_counts)

ST08Q01  ST09Q01
1.0   124380    75969
3.0    29817    10882
5.0    18881     6282

> **ISSUE 4:** 
- Unnormalized scores among previous variable groups.

**Define:** Normalize all scores so that they rate out of 10.

For each score `s` and its max rating `r`, calculate the new normalized score `x` as follows:

$$ \frac{s}{r} = \frac{x}{10} \rightarrow x = \frac{10s}{r}$$

> The reason I picked ***normalization*** for all these scores instead of relying on raw counts is because during the exploratory data visualization phase, it'll be hard to make comparisons between scores when there is no base relativity. 

- for **groups 1 and 2**, their highest score `r` is 4.
- for **groups 3 and 4**, their highest score `r` is 5.

### Code

In [84]:
#isolate group scores
groups1_2 = np.concatenate([group1, group2]).ravel()
groups3_4 = np.concatenate([group3, group4]).ravel()

In [85]:
#check that groups 1 and 2 highest value is 4
pisa2012_clean[groups1_2].max().max()

4.0

In [86]:
#check that groups 3 and 4 highest value is 5
pisa2012_clean[groups3_4].max().max()

5.0

In [87]:
#normalize groups 1 and 2
pisa2012_clean[groups1_2] = pisa2012_clean[groups1_2].apply(lambda s: (10 * s) / 4)

In [88]:
#normalize groups 3 and 4
pisa2012_clean[groups3_4] = pisa2012_clean[groups3_4].apply(lambda s: (10 * s) / 5)

### Test

In [89]:
#isolate all groups
all_groups = np.concatenate([group1, group2, group3, group4]).ravel()

In [90]:
#there should be 67 columns (76 minus 9 from student_info)
pisa2012_clean[all_groups].shape

(485490, 67)

In [91]:
pisa2012_clean[all_groups].sample(5)

ST83Q01  ST83Q02  ST83Q03  ST83Q04  ST86Q01  ST86Q02  ST86Q03  \
87945      10.0     10.0     10.0     10.0     10.0      7.5      7.5   
216898     10.0      7.5      7.5      7.5     10.0      7.5     10.0   
334372      NaN      NaN      NaN      NaN      NaN      NaN      NaN   
407127     10.0      7.5     10.0      7.5      7.5      5.0      7.5   
140894     10.0      5.0      7.5      7.5      5.0      7.5      5.0   

        ST86Q04  ST86Q05  ST87Q01  ...  ST79Q06  ST79Q08  ST79Q15  ST93Q01  \
87945      10.0     10.0     10.0  ...     10.0     10.0     10.0     10.0   
216898     10.0      7.5      7.5  ...     10.0      7.5      7.5      6.0   
334372      NaN      NaN      NaN  ...      NaN      NaN      NaN      6.0   
407127      7.5     10.0     10.0  ...     10.0     10.0     10.0      NaN   
140894      5.0      5.0      7.5  ...     10.0      5.0      5.0      NaN   

        ST93Q03  ST93Q04  ST93Q06  ST93Q07  ST08Q01  ST09Q01  
87945      10.0      8.0      8.0      6.0      NaN      NaN  
216898      6.0      8.0      6.0      6.0      2.0      NaN  
334372      2.0      6.0      8.0      8.0      NaN      NaN  
407127      NaN      NaN      NaN      NaN      NaN      NaN  
140894      NaN      NaN      NaN      NaN      NaN      NaN  

[5 rows x 67 columns]

In [92]:
#check that no string value exists
pisa2012_clean[all_groups].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 67 columns):
ST83Q01    313505 non-null float64
ST83Q02    313112 non-null float64
ST83Q03    312943 non-null float64
ST83Q04    312945 non-null float64
ST86Q01    313223 non-null float64
ST86Q02    312591 non-null float64
ST86Q03    312188 non-null float64
ST86Q04    312294 non-null float64
ST86Q05    311904 non-null float64
ST87Q01    311776 non-null float64
ST87Q02    312138 non-null float64
ST87Q03    310821 non-null float64
ST87Q04    310998 non-null float64
ST87Q05    310587 non-null float64
ST87Q06    310952 non-null float64
ST87Q07    310281 non-null float64
ST87Q08    310735 non-null float64
ST87Q09    311101 non-null float64
ST23Q01    13730 non-null float64
ST23Q03    13497 non-null float64
ST23Q05    13450 non-null float64
ST23Q07    13411 non-null float64
ST24Q01    13457 non-null float64
ST24Q02    13351 non-null float64
ST24Q03    13281 non-null float64
ST46Q01    313898 non

> **ISSUE 5:** 
- Under mother_occupation and father_occupation, 'Missing' is in place of 'NaN.'

**Define:** Replace all 'Missing' strings with 'NaN' under parent occupation fields.

### Code

In [98]:
pisa2012_clean[['OCOD1', 'OCOD2']] = pisa2012_clean[['OCOD1', 'OCOD2']].replace('Missing', np.NaN)

### Test

In [110]:
#check if 'Missing' exists in either parent occupation
(pisa2012_clean['OCOD1'] == 'Missing').any() or (pisa2012_clean['OCOD2'] == 'Missing').any()

False

> **ISSUE 5:** 
- Student ID not explicitly defined as a string.

**Define:** Turn Student ID field into a string datatype.

### Code

In [113]:
pisa2012_clean['STIDSTD'] = pisa2012_clean['STIDSTD'].astype(str)

### Test

In [114]:
pisa2012_clean.STIDSTD.dtype

dtype('O')

## Tidiness
> **ISSUE 1:** Abundance of unnecessary variables besides the ones isolated in pre-assessment stage.

**Define:** 

1. engineer new columns: `teacher_helpfulness`, `belonging`, `acculturation`, `work_ethic`, `future_optimism`, and `self_esteem`. The values for these columns will be the **average score** of all associated questions.

2. under the `student_info` group, rename these columns accordingly:
    - STIDSTD : `student_ID`
    - CNT : `country_code`
    - ST04Q01 : `gender`
    - PV1MATH : `math_score`
    - PV1READ : `reading_score`
    - PV1SCIE : `science_score`
    - OCOD1 : `mother_occupation`
    - OCOD2 : `father_occupation`
    - ST22Q01 : `mother_immigrant`
    
3. keep only the above 15 columns.

## 1. Engineer new columns

### Code


In [115]:
new_column_labels = {'teacher_helpfulness' : teacher_helpfulness,
                     'belonging' : belonging,
                     'acculturation' : acculturation,
                     'work_ethic' : work_ethic,
                     'future_optimism' : future_optimism,
                     'self_esteem' : self_esteem}

In [116]:
#engineer new columns
for key, cols in new_column_labels.items():
    pisa2012_clean[key] = pisa2012_clean[cols].mean(axis = 1)

### Test

In [117]:
groups_renamed = list(new_column_labels.keys())
pisa2012_clean[groups_renamed].head()

teacher_helpfulness  belonging  acculturation  work_ethic  future_optimism  \
0             6.785714   6.111111            NaN    6.964286             6.75   
1                  NaN        NaN            NaN    7.900000             5.00   
2                  NaN        NaN            NaN    9.250000            10.00   
3            10.000000        NaN            NaN         NaN              NaN   
4             8.035714  10.000000            NaN    8.642857             9.75   

   self_esteem  
0     6.818182  
1          NaN  
2          NaN  
3     6.500000  
4     7.500000

## 2. Rename `student_info` fields

### Code

In [118]:
student_info_labels = {'STIDSTD' : 'student_id',
                        'CNT' : 'country_code',
                        'ST04Q01' : 'gender',
                        'PV1MATH' : 'math_score',
                        'PV1READ' : 'reading_score',
                        'PV1SCIE' : 'science_score',
                        'OCOD1' : 'mother_occupation',
                        'OCOD2' : 'father_occupation',
                        'ST22Q01' : 'mother_immigrant'}

In [119]:
#rename student_info labels
pisa2012_clean = pisa2012_clean.rename(columns = student_info_labels)

### Test

In [120]:
#verify renaming has been done correctly
student_info_renamed = list(student_info_labels.values())
pisa2012_clean[student_info_renamed].info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 9 columns):
student_id           485490 non-null object
country_code         485490 non-null object
gender               485490 non-null object
math_score           485490 non-null float64
reading_score        485490 non-null float64
science_score        485490 non-null float64
mother_occupation    456843 non-null object
father_occupation    446377 non-null object
mother_immigrant     40721 non-null object
dtypes: float64(3), object(6)
memory usage: 33.3+ MB


## 3. Keep only the 15 specified columns

### Code

In [121]:
#specify columns to keep
keep = np.concatenate([student_info_renamed, groups_renamed]).ravel()
pisa2012_clean = pisa2012_clean[keep]

### Test

In [122]:
#verify we have 15 columns
pisa2012_clean.columns.shape[0]

15

In [123]:
#check data types
pisa2012_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 485490 entries, 0 to 485489
Data columns (total 15 columns):
student_id             485490 non-null object
country_code           485490 non-null object
gender                 485490 non-null object
math_score             485490 non-null float64
reading_score          485490 non-null float64
science_score          485490 non-null float64
mother_occupation      456843 non-null object
father_occupation      446377 non-null object
mother_immigrant       40721 non-null object
teacher_helpfulness    317870 non-null float64
belonging              313548 non-null float64
acculturation          14201 non-null float64
work_ethic             391137 non-null float64
future_optimism        472336 non-null float64
self_esteem            317427 non-null float64
dtypes: float64(9), object(6)
memory usage: 55.6+ MB


# STORE

In [124]:
pisa2012_clean.to_csv('./data/pisa2012_clean.csv', encoding = 'utf-8', index = False)